controlplane/kubeadm/internal/workload_cluster_coredns.go

```diff
@@ -273,6 +278,12 @@ func (w *Workload) updateCoreDNSClusterRole(ctx context.Context, kubernetesVersi
 		return nil
 	}
 
+	// Do nothing for Kubernetes > 1.22 and coreDNS Version > 1.8.1.
+	if kubernetesVersion.GTE(semver.Version{Major: 1, Minor: 23, Patch: 0}) &&
+		targetCoreDNSVersion.GTE(semver.Version{Major: 1, Minor: 8, Patch: 1}) {
```

Comment:
I think for this if we have to use the sourceCoreDNSVersion so we know that we previously upgraded to a CoreDNS version >= v1.8.1